packages/integrations/cloudflare/src/index.ts

```diff
@@ -90,33 +93,91 @@ export default function createIntegration(args?: Options): AstroIntegration {
 					vite.ssr.target = 'webworker';
 				}
 			},
+			'astro:build:ssr': ({ manifest, entryPoints }) => {
+				// NOTE: monkey-patch because entryPoints URL does not respect `build.server`
+				entryPoints.forEach((value, key) => {
+					const oldFilePath = fileURLToPath(value);
+					const relativeFilePath = relative(fileURLToPath(_config.outDir), oldFilePath);
+					const newFilePath = join(fileURLToPath(_buildConfig.server), relativeFilePath);
+					const newEntryUrl = pathToFileURL(newFilePath);
+					_entryPoints.set(key, newEntryUrl);
+				});
+			},
 			'astro:build:done': async ({ pages, routes, dir }) => {
-				const entryPath = fileURLToPath(new URL(_buildConfig.serverEntry, _buildConfig.server));
 				const entryUrl = new URL(_buildConfig.serverEntry, _config.outDir);
 				const buildPath = fileURLToPath(entryUrl);
 				// A URL for the final build path after renaming
 				const finalBuildUrl = pathToFileURL(buildPath.replace(/\.mjs$/, '.js'));
+				const functionsUrl = new URL('functions/', _config.root);
 
-				await esbuild.build({
-					target: 'es2020',
-					platform: 'browser',
-					conditions: ['workerd', 'worker', 'browser'],
-					entryPoints: [entryPath],
-					outfile: buildPath,
-					allowOverwrite: true,
-					format: 'esm',
-					bundle: true,
-					minify: _config.vite?.build?.minify !== false,
-					banner: {
-						js: SHIM,
-					},
-					logOverride: {
-						'ignored-bare-import': 'silent',
-					},
-				});
+				if (isModeDirectory) {
+					await fs.promises.mkdir(functionsUrl, { recursive: true });
+				}
+
+				if (isModeDirectory && _buildConfig.split) {
+					const entryPaths = [..._entryPoints.values()].map((entry) => fileURLToPath(entry));
+					const outputDir = fileURLToPath(new URL('.astro', _buildConfig.server));
+
+					// NOTE: AFAIK, esbuild keeps the order of the entryPoints array
+					const { outputFiles } = await esbuild.build({
+						target: 'es2020',
+						platform: 'browser',
+						conditions: ['workerd', 'worker', 'browser'],
+						entryPoints: entryPaths,
+						outdir: outputDir,
+						allowOverwrite: true,
+						format: 'esm',
+						bundle: true,
+						minify: _config.vite?.build?.minify !== false,
+						banner: {
+							js: SHIM,
+						},
+						logOverride: {
+							'ignored-bare-import': 'silent',
+						},
+						write: false,
+					});
+
+					for (const [index, outputFile] of outputFiles.entries()) {
+						const fileName = [..._entryPoints.entries()][index][0].component.replace('src/pages/', '').replace('.astro', '.js');
+						const fileUrl = new URL(fileName, functionsUrl)
+						const newFileDir = dirname(fileURLToPath(fileUrl));
+						if (!fs.existsSync(newFileDir)) {
+							fs.mkdirSync(newFileDir, { recursive: true });
+						}
+						await fs.promises.writeFile(fileUrl, outputFile.contents);
+					}
```

Comment:
This is definitely the easier API, instead of moving files. However I have not done perf benchmarks, but should be fast enough. If not, we can always go back to the old approach. let me know!